File: quick-start/python/RedShiftServerlessDataAPI.py
import time
import traceback
import boto3
import logging
from collections import OrderedDict

logger = logging.getLogger()
logger.setLevel(logging.INFO)


def lambda_handler(event, context):
    # input parameters passed from the caller event
    # Amazon Redshift Serverless Workgroupname
    redshift_workgroup_name=event['redshift_workgroup_name']
    # database name for the Amazon Redshift serverless instance
    redshift_database_name = event['redshift_database']
    # IAM Role of Amazon Redshift sererless instance having access to S3
    redshift_iam_role = event['redshift_iam_role']
    # run_type can be either asynchronous or synchronous; try tweaking based on your requirement
    run_type = event['run_type']

    sql_statements = OrderedDict()
    res = OrderedDict()

    if run_type != "synchronous" and run_type != "asynchronous":
        raise Exception(
            "Invalid Event run_type. \n run_type has to be synchronous or asynchronous.")

    isSynchronous = True if run_type == "synchronous" else False

    # initiate redshift-data redshift_data_api_client in boto3
    redshift_data_api_client = boto3.client('redshift-data')

    sql_statements['CREATE'] = "CREATE TABLE IF NOT EXISTS public.region (" + \
        "R_REGIONKEY bigint NOT NULL," + \
        "R_NAME varchar(25)," + \
        "R_COMMENT varchar(152))" + \
        "diststyle all;"
    sql_statements['COPY'] = "COPY public.region FROM 's3://redshift-immersionday-labs/data/region/region.tbl.lzo' " + \
        "iam_role '" + redshift_iam_role + \
        "' region 'us-west-2' lzop delimiter '|' COMPUPDATE PRESET;"
    sql_statements['UPDATE'] = "UPDATE public.region set r_regionkey = 5 where r_name ='AFRICA';"
    sql_statements['DELETE'] = "DELETE From public.region where r_name = 'MIDDLE EAST';"
    sql_statements['SELECT'] = "SELECT r_regionkey,r_name from public.region;"
    logger.info("Running sql queries in {} mode!\n".format(run_type))

    try:
        for command, query in sql_statements.items():
            logging.info("Example of {} command :".format(command))
            res[command + " STATUS: "] = execute_sql_data_api(redshift_data_api_client, redshift_database_name, command, query,
                                                              redshift_workgroup_name, isSynchronous)

    except Exception as e:
        raise Exception(str(e) + "\n" + traceback.format_exc())
    return res


def execute_sql_data_api(redshift_data_api_client, redshift_database_name, command, query, redshift_workgroup_name, isSynchronous):

    MAX_WAIT_CYCLES = 20
    attempts = 0
    # Calling Redshift Data API with executeStatement()
    res = redshift_data_api_client.execute_statement(
        Database=redshift_database_name, WorkgroupName=redshift_workgroup_name, Sql=query)
    query_id = res["Id"]
    desc = redshift_data_api_client.describe_statement(Id=query_id)
    query_status = desc["Status"]
    logger.info(
        "Query status: {} .... for query-->{}".format(query_status, query))
    done = False

    # Wait until query is finished or max cycles limit has been reached.
    while not done and isSynchronous and attempts < MAX_WAIT_CYCLES:
        attempts += 1
        time.sleep(1)
        desc = redshift_data_api_client.describe_statement(Id=query_id)
        query_status = desc["Status"]

        if query_status == "FAILED":
            raise Exception('SQL query failed:' +
                            query_id + ": " + desc["Error"])

        elif query_status == "FINISHED":
            logger.info("query status is: {} for query id: {} and command: {}".format(
                query_status, query_id, command))
            done = True
            # print result if there is a result (typically from Select statement)
            if desc['HasResultSet']:
                response = redshift_data_api_client.get_statement_result(
                    Id=query_id)
                logger.info(
                    "Printing response of {} query --> {}".format(command, response['Records']))
        else:
            logger.info(
                "Current working... query status is: {} ".format(query_status))

    # Timeout Precaution
    if done == False and attempts >= MAX_WAIT_CYCLES and isSynchronous:
        logger.info("Limit for MAX_WAIT_CYCLES has been reached before the query was able to finish. We have exited out of the while-loop. You may increase the limit accordingly. \n")
        raise Exception("query status is: {} for query id: {} and command: {}".format(
            query_status, query_id, command))

    return query_status

File: quick-start/javascript/RedshiftServerlessDataAPI.js
'use strict';

const AWS = require('aws-sdk');

// If using Lambda layers, uncomment the line below to import the module
// import { RedshiftData } from "@aws-sdk/client-redshift-data";

exports.handler = async (event) => {
  const {
    // input parameters passed from the caller event
	// Amazon Redshift Serverless Workgroupname
	redshift_workgroup_name: redshiftWorkgroupName,
    // database name for the Amazon Redshift cluster
    redshift_database: redshiftDatabaseName,
    // IAM Role of Amazon Redshift cluster having access to S3
    redshift_iam_role: redshiftIAMRole,
    // run_type can be either asynchronous or synchronous; try tweaking based on your requirement
    run_type: runType
  } = event;

  const responses = new Map();

  if (runType !== 'synchronous' && runType !== 'asynchronous') {
    throw new Error('Invalid Event run_type. \n run_type has to be synchronous or asynchronous.');
  }

  const isSynchronous = (runType === 'synchronous');

  // returns a Map(), ES6+ preserves insertion ordering
  const sqlStatements = await populateSqlStatementSet(redshiftIAMRole);

  console.log(`Running sql queries in ${runType} mode. \n`);

  // Initiate RedshiftData client
  const redshiftDataApiClient = new AWS.RedshiftData({ region: 'us-east-1' });

  // If using Lambda layers, uncomment the line below
  // const redshiftDataApiClient = new RedshiftData({ region: "us-east-1" });

  for (const [command, query] of sqlStatements.entries()) {
    console.log(`Example of ${command} mode.`);
    const res = await executeSqlDataApi(redshiftDataApiClient, redshiftWorkgroupName, redshiftDatabaseName, command, query, isSynchronous);

    responses.set(`${command} STATUS: `, res);
  }

  return {
    statusCode: 200,
    body: [...responses.entries()]
  };
};

const executeSqlDataApi = async (redshiftDataApiClient, redshiftWorkgroupName, redshiftDatabaseName, command, query, isSynchronous) => {
  let queryId = '';

  const executeStatementInput = {
    WorkgroupName: redshiftWorkgroupName,
    Database: redshiftDatabaseName,
    Sql: query
  };

  // Calling Redshift Data API with executeStatement()
  await redshiftDataApiClient.executeStatement(executeStatementInput).promise()
    .then((response) => {
      queryId = response.Id;
    })
    .catch((error) => {
      console.log('ExecuteStatement has failed.');
      throw new Error(error);
    });

  let { Status: queryStatus } = await getDescribeStatement(redshiftDataApiClient, queryId);

  console.log(`Executed command: ${command} | Query Status: ${queryStatus} | QueryId: ${queryId}`);

  if (isSynchronous) {
    queryStatus = await executeSynchronousWait(redshiftDataApiClient, queryStatus, queryId, command);
  }

  return queryStatus;
};

const executeSynchronousWait = async (redshiftDataApiClient, queryStatus, queryId, command) => {
  let attempts = 0;
  const MAX_WAIT_CYCLES = 20;
  let describeStatementInfo = {};

  // Wait until query is finished or max cycles limit has been reached.
  while (attempts < MAX_WAIT_CYCLES) {
    attempts++;
    await sleep(1);

    ({ Status: queryStatus, ...describeStatementInfo } = await getDescribeStatement(redshiftDataApiClient, queryId));

    if (queryStatus === 'FAILED') {
      throw new Error(`SQL query failed: ${queryId}: \n Error: ${describeStatementInfo.Error}`);
    } else if (queryStatus === 'FINISHED') {
      console.log(`Query status is: ${queryStatus} for query id: ${queryId} and command: ${command}`);

      // Print query response if available (typically from Select SQL statements)
      if (describeStatementInfo.HasResultSet) {
        await redshiftDataApiClient.getStatementResult({ Id: queryId }).promise()
          .then((statementResult) => {
            console.log(`Printing response for query: ${command} --> ${JSON.stringify(statementResult.Records)}`);
          })
          .catch((error) => {
            console.log('GetStatementResult has failed.');
            throw new Error(error);
          });
      }

      break;
    } else {
      console.log(`Currently working... query status is ${queryStatus}`);
    }

    if (attempts >= MAX_WAIT_CYCLES) {
      throw new Error(`Limit for MAX_WAIT_CYCLES has been reached before the query was able to finish. We have exited out of the while-loop. You may increase the limit accordingly. \n Query status is: %s for query id: ${queryId} and command: ${command}`);
    }
  }
  return queryStatus;
};

const getDescribeStatement = async (redshiftDataApiClient, queryId) => redshiftDataApiClient
  .describeStatement({ Id: queryId })
  .promise()
  .then(response => {
    return response;
  })
  .catch((error) => {
    console.log('DescribeStatement has failed.');
    throw new Error(error);
  });

const populateSqlStatementSet = async (redshiftIAMRole) => {
  const sqlStatements = new Map();
  sqlStatements.set('CREATE', 'CREATE TABLE IF NOT EXISTS public.region (\n' + '  R_REGIONKEY bigint NOT NULL,\n' +
                                '  R_NAME varchar(25),\n' + '  R_COMMENT varchar(152))\n' + 'diststyle all;');

  sqlStatements.set('COPY', "COPY region FROM 's3://redshift-immersionday-labs/data/region/region.tbl.lzo'\n" +
                            "iam_role '" + redshiftIAMRole + "' \n" + "region 'us-west-2' lzop delimiter '|' COMPUPDATE PRESET;");

  sqlStatements.set('UPDATE', "UPDATE public.region SET r_regionkey= 5 WHERE r_name ='AFRICA';");

  sqlStatements.set('DELETE', "DELETE FROM public.region where r_name = 'MIDDLE EAST';");

  sqlStatements.set('SELECT', 'SELECT r_regionkey, r_name FROM public.region;');
  return sqlStatements;
};

const sleep = (seconds) => {
  return new Promise(resolve => setTimeout(resolve, seconds * 1000));
};
